pkg/ipamd/datastore/data_store.go

```diff
@@ -655,67 +563,59 @@ func (ds *DataStore) AssignPodIPv4Address(ipamKey IPAMKey) (ipv4address string,
 
 	ds.log.Debugf("AssignIPv4Address: IP address pool stats: total: %d, assigned %d", ds.total, ds.assigned)
 
-	if eni, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
+	if eni, _, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
 		ds.log.Infof("AssignPodIPv4Address: duplicate pod assign for sandbox %s", ipamKey)
 		return addr.Address, eni.DeviceNumber, nil
 	}
 
 	for _, eni := range ds.eniPool {
-		ds.log.Debugf("Checking for eni %v", eni)
-		if !eni.IsPDEnabled {
-			for _, addr := range eni.IPv4Addresses {
-				if !addr.Assigned() && !addr.inCoolingPeriod() {
-					ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
-					if err := ds.writeBackingStoreUnsafe(); err != nil {
-						ds.log.Warnf("Failed to update backing store: %v", err)
-						// Important! Unwind assignment
-						ds.unassignPodIPv4AddressUnsafe(addr)
-						return "", -1, err
-					}
-
-					return addr.Address, eni.DeviceNumber, nil
+		for _, availableCidr := range eni.AvailableIPv4Cidrs {
+			var addr *AddressInfo
+			var strPrivateIPv4 string
+			var err error
+
+			if ds.isPDEnabled && availableCidr.IsPrefix {
+				strPrivateIPv4, err = ds.getFreeIPv4AddrfromCidr(availableCidr)
```

Comment:
why for cidr, we use `getFreeIPv4AddrfromCidr(availableCidr)`
but for sip, we didn't use it. 
the place where we check for assigned/cooldown differs between this two code path (but ideally it can be treated same way)

Reply:
With PD, we try to get a free unused IP/32 from /28 hence I abstracted that in a function but whereas for SIP we already have a /32. As you suggested(next comment), in the following PR, I will create a function under CidrInfo to allocate a free IP.